daemon/api_download_test.go

```diff
@@ -0,0 +1,192 @@
+	"github.com/snapcore/snapd/store/storetest"
+)
+
+type fakeStore struct{}
+
+var _ = check.Suite(&snapDownloadSuite{})
+
+type snapDownloadSuite struct {
+	storetest.Store
+	d *daemon.Daemon
+
+	snaps []string
+}
+
+func (s *snapDownloadSuite) SetUpTest(c *check.C) {
+	s.snaps = nil
+
+	o := overlord.Mock()
+	s.d = daemon.NewWithOverlord(o)
+
+	st := o.State()
+	st.Lock()
+	defer st.Unlock()
+	snapstate.ReplaceStore(st, s)
+	dirs.SetRootDir(c.MkDir())
+}
+
+func (s *snapDownloadSuite) SnapInfo(spec store.SnapSpec, user *auth.UserState) (*snap.Info, error) {
+	if spec.Name == "bar" {
+		return &snap.Info{
+			DownloadInfo: snap.DownloadInfo{
+				Size:            12,
+				AnonDownloadURL: "http://localhost/bar",
+			},
+		}, nil
+	}
+	if spec.Name == "foo" {
+		return &snap.Info{
+			DownloadInfo: snap.DownloadInfo{
+				Size:            100,
+				AnonDownloadURL: "http://localhost/foo",
+			},
+		}, nil
+	}
+	return nil, store.ErrSnapNotFound
+}
+
+func (s *snapDownloadSuite) DownloadStream(ctx context.Context, name string, downloadInfo *snap.DownloadInfo, user *auth.UserState) (io.ReadCloser, error) {
+	if name == "bar" {
+		return ioutil.NopCloser(bytes.NewReader([]byte("SNAP"))), nil
+	} else {
+		return nil, fmt.Errorf("unexpected error")
+	}
+}
+
+func (s *snapDownloadSuite) TestDownloadSnapErrors(c *check.C) {
+	type scenario struct {
+		dataJSON string
+		status   int
+		err      string
+	}
+
+	for _, scen := range []scenario{
+		{
+			dataJSON: `{"action": "download"}`,
+			status:   400,
+			err:      "download operation requires one snap name",
+		},
+		{
+			dataJSON: `{"action": "foo", "snaps": ["foo"]}`,
+			status:   400,
+			err:      `unknown download operation "foo"`,
+		},
+		{
+			dataJSON: `{"snaps": ["foo"]}`,
+			status:   400,
+			err:      `download operation requires action`,
+		},
+		{
+			dataJSON: `{"action": "foo", "snaps": ["foo", "bar"]}`,
+			status:   400,
+			err:      `download operation supports only one snap`,
+		},
+		{
+			dataJSON: `{"}`,
+			status:   400,
+			err:      `cannot decode request body into download operation: unexpected EOF`,
+		},
+	} {
+		var err error
+		data := []byte(scen.dataJSON)
+
+		req, err := http.NewRequest("POST", "/v2/download", bytes.NewBuffer(data))
+		c.Assert(err, check.IsNil)
+		rsp := daemon.PostSnapDownload(daemon.SnapDownloadCmd, req, nil)
+
+		c.Assert(rsp.(*daemon.Resp).Status, check.Equals, scen.status)
+		if scen.err == "" {
+			c.Errorf("error was expected")
+		}
+		result := rsp.(*daemon.Resp).Result
+		c.Check(result.(*daemon.ErrorResult).Message, check.Matches, scen.err)
+	}
+}
+
+func (s *snapDownloadSuite) TestStreamOneSnap(c *check.C) {
+
+	type scenario struct {
+		dataJSON string
+		status   int
+		err      string
+	}
+
+	for _, s := range []scenario{
+		{
+			dataJSON: `{"action": "download", "snaps": ["doom"]}`,
+			status:   404,
+			err:      "snap not found",
+		},
+		{
+			dataJSON: `{"action": "download", "snaps": ["foo"]}`,
```

Comment:
I would rename "foo" to something like "download-error-trigger-snap" to make reading a bit easier. I stumbled a bit reading this as the dataJSON that triggers the error looks almost the same as the one that does not, of course the snap is different but "foo" looks innocent :)